### PR TITLE
Support RCEA with independent states

### LINKED ISSUE #3769: Allow `RequestContextExporter`s to have independent states
Sometimes, users (or maybe just me 😄 ) may want to leave logs with certain mdc properties for specific logs only.

Examples:
1) Record a single log with specific mdc's per request for KPI aggregation
2) MDC properties may become quite large, and users may not want to record all mdc values for every debug log.

I wonder if we could specify an option in `RequestContextExporter` which creates a clean MDC map.

Code for reproduction
```
    @Test
    void multipleExporters() {
        final ServiceRequestContext ctx = ServiceRequestContext.of(HttpRequest.of(HttpMethod.GET, "/"));
        final RequestContextExporter exporter1 =
                RequestContextExporter.builder()
                                      .attr("attrs.attr1", ATTR1)
                                      .build();
        final RequestContextExporter exporter2 =
                RequestContextExporter.builder()
                                      .attr("attrs.attr2", ATTR2)
                                      .build();

        ctx.setAttr(ATTR1, "foo");
        assertThat(exporter1.export(ctx)).containsOnlyKeys("attrs.attr1");

        ctx.setAttr(ATTR2, "bar");
        assertThat(exporter2.export(ctx)).containsOnlyKeys("attrs.attr2");
    }
```

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

- `RCEA`s now only export the specified `mdc`s

Modifications:

- Modify the `STATE` `AttributeKey` used in `RequestContextExporter` to be generated per instance

Result:

- Closes #3769

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/common/logging/RequestContextExporterTest.java [hunk 1]
```diff
@@ -165,6 +165,27 @@ void customExportKey() {
...
+        final RequestContextExporter exporter1 =
+                RequestContextExporter.builder()
+                                      .name("exporter1")
+                                      .attr("attrs.attr1", ATTR1)
+                                      .build();
+        final RequestContextExporter exporter2 =
+                RequestContextExporter.builder()
+                                      .name("exporter2")
+                                      .attr("attrs.attr2", ATTR2)
+                                      .build();
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/common/logging/RequestContextExporterTest.java:180]: If `exporter1` and `exporter2` have the same name, this test seems to fail.
How about appending a unique number to `AttributeKey`.
```java
private static final AtomicLong counter = new AtomicLong();
stateAttributeKey = AttributeKey.valueOf(RequestContextExporter.class, name + counter.getAndIncrement() + "_STATE");
```

File: core/src/main/java/com/linecorp/armeria/common/logging/RequestContextExporterBuilder.java [hunk 2]
```diff
@@ -40,9 +40,19 @@
...
     RequestContextExporterBuilder() {}
 
+    /**
+     * Sets the specified {@code name} for the {@link RequestContextExporter}.
+     */
+    public RequestContextExporterBuilder name(String name) {
+        requireNonNull(name, "name");
+        this.name = name;
+        return this;
+    }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RequestContextExporterBuilder.java:54]: Is there a case where a user would want to share the same attribute key? If not, we could remove this method and always make sure each `RequestContextExporter` has its own attribute key.

**Comment by jrhee17** [comment]
> If exporter1 and exporter2 have the same name, this test seems to fail.
How about appending a unique number to AttributeKey.

> Is there a case where a user would want to share the same attribute key? If not, we could remove this method and always make sure each RequestContextExporter has its own attribute key.

At least I can't think of any use cases where users would want to share the same key - updated!

**Review by ikhoon — APPROVED** [review]
Thanks! @jrhee17

File: logback/src/test/java/com/linecorp/armeria/common/logback/RequestContextExportingAppenderTest.java [hunk 3]
```diff
@@ -253,6 +253,43 @@ void testXmlConfigExportPrefix() throws Exception {
...
+
+            assertThat(rcea1.exporter().export(ctx)).containsOnlyKeys("remote.ip");
+            assertThat(rcea2.exporter().export(ctx)).containsOnlyKeys("remote.port");
+        } finally {
+            // Revert to the original configuration.
+            final JoranConfigurator configurator = new JoranConfigurator();
+            configurator.setContext(context);
+            context.reset();
+
+            configurator.doConfigure(getClass().getResource("/logback-test.xml"));
```
Comment (minwoox) [inline logback/src/test/java/com/linecorp/armeria/common/logback/RequestContextExportingAppenderTest.java:289]: How about extracting this into a method to dedup?

File: core/src/test/java/com/linecorp/armeria/common/logging/RequestContextExporterTest.java [hunk 4]
```diff
@@ -165,6 +165,25 @@ void customExportKey() {
...
+        final RequestContextExporter exporter2 =
+                RequestContextExporter.builder()
+                                      .attr("attrs.attr2", ATTR2)
+                                      .build();
+
+        ctx.setAttr(ATTR1, "foo");
+        assertThat(exporter1.export(ctx)).containsOnlyKeys("attrs.attr1");
+
+        ctx.setAttr(ATTR2, "bar");
+        assertThat(exporter2.export(ctx)).containsOnlyKeys("attrs.attr2");
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/common/logging/RequestContextExporterTest.java:184]: nit: How about:
```suggestion
        ctx.setAttr(ATTR1, "foo");
        ctx.setAttr(ATTR2, "bar");
        
        assertThat(exporter1.export(ctx)).containsOnlyKeys("attrs.attr1");     
        assertThat(exporter2.export(ctx)).containsOnlyKeys("attrs.attr2");
```

**Review by minwoox — APPROVED** [review]
Left two nits.
Thanks! 😄

**Review by trustin — APPROVED** [review]
Thanks, @jrhee17 !

**Pull request merged**